pkg/runctl/controller.go

```diff
@@ -412,11 +410,47 @@ func (c *Controller) syncHandler(key string) error {
 		if err = c.changeAndCommitStateAndMeter(ctx, pipelineRun, api.StateWaiting, metav1.Now()); err != nil {
 			return err
 		}
+
 	case api.StateWaiting:
 		run, err := runManager.GetRun(ctx, pipelineRun)
 		if err != nil {
 			return c.onGetRunError(ctx, pipelineRunAPIObj, pipelineRun, err, api.StateCleaning, api.ResultErrorInfra, "waiting failed")
 		}
+		// the configuration should be loaded once per sync to avoid inconsistencies
+		// in case of concurrent configuration changes
+		pipelineRunsConfig, err := c.loadPipelineRunsConfig(ctx)
+		if err != nil {
+			return c.onGetRunError(ctx, pipelineRunAPIObj, pipelineRun, err, api.StateCleaning, api.ResultErrorInfra, "failed to load configuration for pipeline runs")
+		}
+
+		// Check for wait timeout
+		startTime := pipelineRun.GetStatus().StateDetails.StartedAt
+		timeout := pipelineRunsConfig.TimeoutWait
+		if startTime.Add(timeout.Duration).Before(time.Now()) {
+			return c.handleResultError(ctx, pipelineRun, api.ResultErrorInfra, "waiting failed", fmt.Errorf("wait timeout reached"))
+		}
+
+		if run == nil {
+			if err = runManager.Start(ctx, pipelineRun, pipelineRunsConfig); err != nil {
+				c.recorder.Event(pipelineRunAPIObj, corev1.EventTypeWarning, api.EventReasonPreparingFailed, err.Error())
+				resultClass := serrors.GetClass(err)
+				// In case we have a result we can cleanup. Otherwise we retry in the next iteration.
+				if resultClass != api.ResultUndefined {
+					return c.handleResultError(ctx, pipelineRun, resultClass, "waiting failed", err)
+				}
+				return err
+			}
+			return nil
+		}
+
+		if run.IsRestartable() {
+			c.recorder.Event(pipelineRunAPIObj, corev1.EventTypeWarning, api.EventReasonWaitingFailed, "restarting")
+			if err = runManager.DeleteRun(ctx, pipelineRun); err != nil {
+				return c.handleResultError(ctx, pipelineRun, api.ResultErrorInfra, "restart failed", err)
```

Comment:
actually the deletion before the restart failed... should this info be included in the message?

Reply:
See https://github.com/SAP/stewardci-core/pull/345/commits/2cc1b5d00eadebbed52f872855c7298d7e3f6481